File: 2025/python/day6.py
#!/usr/bin/env python3

import operator
from aoc import *
from functools import reduce


def part1(input: str):
    problems = []
    lines = input.splitlines()
    for line in lines[:-1]:
        row = line.split()
        if not problems:
            problems = [[] for _ in range(len(row))]
        for i, v in enumerate(row):
            problems[i].append(int(v))
    total = 0
    for i, op in enumerate(lines[-1].split()):
        problem = problems[i]
        match op:
            case '+':
                total += reduce(operator.add, problem)
            case '*':
                total += reduce(operator.mul, problem) 
            
    return total
    

def part2(input: str):
    lines = input.splitlines()
    total = 0
    op = '+'
    answer = 0
    for i, c in enumerate(lines[-1]):
        if c != ' ':
            total += answer
            op = c
            if c == '*':
                answer = 1
            else:
                answer = 0
        num = None
        for line in lines[:-1]:
            if line[i] != ' ':
                if num is None:
                    num = 0
                else:
                    num = num * 10
                num += int(line[i])
        if num is None:
            total += answer
            answer = 0
        else:
            if op == '+':
                answer += num
            else:
                answer *= num
    total += answer
    return total


if __name__ == '__main__':
    main()

File: 2025/go/day6.go
package main

import (
	"github.com/bpa/aoc"
)

func add(a, b int) int {
	return a + b
}

func mul(a, b int) int {
	return a * b
}

func day6Part1(input string) int {
	lines := aoc.Delimited(input)
	total := 0
	end := len(lines) - 1
	for i, op := range lines[end] {
		var f func(int, int) int
		if op == "*" {
			f = mul
		} else {
			f = add
		}
		subtotal := aoc.ToInt(lines[0][i])
		for _, n := range lines[1:end] {
			subtotal = f(subtotal, aoc.ToInt(n[i]))
		}
		total += subtotal
	}
	return total
}

func day6Part2(input string) int {
	lines := aoc.Lines(input)
	end := len(lines) - 1
	total := 0
	subtotal := 0
	var f func(int, int) int
	for i, c := range lines[end] {
		value := 0
		isNumber := false
		for _, n := range lines[:end] {
			x := n[i]
			if x != ' ' {
				isNumber = true
				value *= 10
				value += int(x) - '0'
			}
		}
		if !isNumber {
			total += subtotal
		} else if c == '*' {
			f = mul
			subtotal = value
		} else if c == '+' {
			f = add
			subtotal = value
		} else {
			subtotal = f(subtotal, value)
		}
	}
	return total + subtotal
}

func main() {
	aoc.Main(day6Part1, day6Part2)
}
